### ISSUE TITLE
Allow disabling automatic linking to static libraries

### ISSUE THREAD

**Issue by yodantauber** [coapp/coapp.powershell#15]
When generating a native NuGet package with Write-NuGetPackage, and when specifying my library's LIB files with "lib:" in the autopkg file, any project that consumes the package will automatically try to link with that lib file (because it is added under "AdditionalDependencies").

This behavior is commonly undesirable - many libraries (e.g. Boost, CGAL) have their own smart auto-linking mechanism built it to their headers, by the means of "#pragma comment" for example. When this is the case, adding the library files to AdditionalDependencies may actually cause linkage errors.

I could not find any way to disable this behavior (while still packaging and distributing the LIB files), and I think there should be.

**Comment by yodantauber** [comment]
I should add that when a LIB file is _not_ added to AdditionalDependencies, its containing folder should probably be added to AdditionalLibraryDirectories so that the pragma comments will work.

**Comment by fearthecowboy** [comment]
Good point about the #pragma comment.

To suppress the automatic link to the libs, and just add the folder to the Linker AdditionalLibraryDirectories, try this:

``` c
// you can use any custom name for 'customlib' 
customlib: { 
    // destination attribute dictates where the files end up 
    #destination : ${d_lib};   

    // add-folder attribute adds the folder to the node specified 
    #add-folder : ::nuget.targets.[${condition}].ItemDefinitionGroup.Link.AdditionalLibraryDirectories;

    // all the files as normal, etc
    foo.dll, bar.dll 
};
```

That should let you have your files end up in the appropriate lib folder, but not auto linking (which is enabled in the default lib rule).

FYI, I am working on a significant revision this week to properly support splitting up very large packages into packages with overlays (since we found some library sets were getting stupidly huge). 

I've also juggled the code a bit, and I'll make sure that the implicitly-defined property sheets are visible so you can see how this stuff got put together.

**Comment by yodantauber** [comment]
Thank you for your reply! Your suggestion seems to work for me.
I just have one more issue: I need to specify libs in 4 different combinations of pivots. Right now the only way I've been able to get this to work correctly was to duplicate the entire "customlib:" block, with the "destination" and "add-folder" definitions, under each conditional. Naturally I would like to define them only once, and only repeat the title "customlib:" under each conditional (like is done with "lib:"). Later on I will want to define customlib in a separate file and include it in many AutoPackage scripts.

I tried adding the following block at the top of the file:

```
nuget := {
    files {
        customlib += { 
            #destination : ${d_lib};   
            #add-folder : ::nuget.targets.[${condition}].ItemDefinitionGroup.Link.AdditionalLibraryDirectories;
        }
    }
}
```

but the result of this is that the following unwanted line is added to the targets file (in addition to the desired library directories):

```
<AdditionalLibraryDirectories>$(MSBuildThisFileDirectory)../..//build/native/lib/;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```

Am I doing something wrong here?

**Comment by fearthecowboy** [comment]
Hmm. No, nothing wrong, I should put a check in to see if the folder actually contains anything before executing the #add-folder ... Lemme see what I can do.

**Comment by fearthecowboy** [comment]
Yeah, I found an easy place to do this. It won't do the #add-folder if the folder doesn't actually contain any items, so that superfluous entry won't happen. I should be publishing a build you can play with tomorrow...

**Comment by yodantauber** [comment]
The current development version (2.4.199.0) doesn't add the superfluous entry anymore.
Before we close this issue, do you have any plans to add this "customlib" (probably named as "lib_nolink") to the default script? Or should this remain a true customization? I think it would be useful to others as well.

**Comment by yodantauber** [comment]
Hmm. Actually, this change seems to interfere with nestedInclude: earlier, the root "include" folder would be added as an include folder. Now, only the include/something folder is added as an include folder. This fails `#include "something/file.h"` directives in the code. 

In other words, the "include" folder itself should be added as an include folder when nestedInclude is used.

**Comment by barrylapthorn** [comment]
Firstly, @fearthecowboy - thanks for doing all this, it is really useful.

The tl;dr bit - this thread is the solution to my problem, but can we get @yodantauber 's suggestion of 'lib_nolink' added (and then added to coapp.org's docs)?  I'm using v1.23.521 and I can't see that, so I'm guessing it hasn't been added.  Or I'm blind.  In the meantime the work around from above:

```
files {
    customlib += { 
        #destination : ${d_lib};   
        #add-folder : ::nuget.targets.[${condition}].ItemDefinitionGroup.Link.AdditionalLibraryDirectories;
    }
}
```

does the job.

The longer bit, my problem:  I've been trying to package boost for a corporate system, and boost uses the `#pragma comment(lib, 'lib.lib')` trick to implicitly link libraries, without the developer having to add them to the linker input additional dependencies line.  In my case, the coapp `lib:` rule ends up including every static lib, which is unnecessary, additionally including the boost unit test lib which also includes an `int main` entry point.  Which then causes a `int main has already been defined` error if I add the boost package to a regular (non-unit test exe).

Anyway, it's the same problem as the one in this thread.  

Thanks again.
